File: indexer/etherscan.py
"""
Etherscan API module — pulls full transaction history for any address.

Handles ERC-20 token transfers and native ETH/BNB/MATIC transfers.
All functions are async and paginate automatically (max 10,000 txs per call).
"""

import httpx
import asyncio
from datetime import datetime
from config.chains import CHAINS
from config.tokens import BURN_ADDRESS

# Limit concurrent calls to Etherscan V2 to avoid rate-limiting (5 req/s free tier)
_ETHERSCAN_SEM = asyncio.Semaphore(2)


async def _fetch(url: str, params: dict) -> dict:
    """Low-level async GET with retries. Throttles Etherscan V2 calls."""
    from config.chains import ETHERSCAN_V2
    use_sem = (url == ETHERSCAN_V2)
    async with httpx.AsyncClient(timeout=30, follow_redirects=True) as client:
        for attempt in range(3):
            try:
                if use_sem:
                    async with _ETHERSCAN_SEM:
                        r = await client.get(url, params=params)
                        await asyncio.sleep(0.3)  # 300ms gap between V2 calls
                else:
                    r = await client.get(url, params=params)
                r.raise_for_status()
                return r.json()
            except Exception as e:
                if attempt == 2:
                    raise
                await asyncio.sleep(1.5 ** attempt)
    return {}


def _params(cfg: dict, base: dict) -> dict:
    """Build explorer params — add chainid only for Etherscan V2 endpoints."""
    if cfg.get("explorer_v2", True):
        base["chainid"] = cfg["chain_id"]
    if cfg["explorer_key"]:
        base["apikey"] = cfg["explorer_key"]
    return base


async def fetch_token_transfers(wallet: str, chain: str) -> list[dict]:
    """Fetch all ERC-20 token transfers for a wallet on a chain."""
    cfg = CHAINS[chain]
    data = await _fetch(cfg["explorer"], _params(cfg, {
        "module":  "account",
        "action":  "tokentx",
        "address": wallet,
        "sort":    "asc",
    }))

    if data.get("status") != "1":
        return []

    flows = []
    for tx in data["result"]:
        decimals = int(tx.get("tokenDecimal") or 18)
        flows.append({
            "tx_hash":       tx["hash"],
            "chain":         chain,
            "from_address":  tx["from"].lower(),
            "to_address":    tx["to"].lower(),
            "token":         tx["tokenSymbol"],
            "token_address": tx["contractAddress"].lower(),
            "amount":        int(tx["value"]) / (10 ** decimals),
            "block_number":  int(tx["blockNumber"]),
            "block_time":    datetime.fromtimestamp(int(tx["timeStamp"])),
        })
    return flows


async def fetch_native_transfers(wallet: str, chain: str) -> list[dict]:
    """Fetch all native transfers (ETH/BNB/MATIC) for a wallet on a chain."""
    cfg = CHAINS[chain]
    data = await _fetch(cfg["explorer"], _params(cfg, {
        "module":  "account",
        "action":  "txlist",
        "address": wallet,
        "sort":    "asc",
    }))

    if data.get("status") != "1":
        return []

    flows = []
    for tx in data["result"]:
        value_eth = int(tx["value"]) / 1e18
        if value_eth == 0:
            continue
        flows.append({
            "tx_hash":       tx["hash"],
            "chain":         chain,
            "from_address":  tx["from"].lower(),
            "to_address":    tx.get("to", "").lower(),
            "token":         CHAINS[chain]["native"],
            "token_address": "native",
            "amount":        value_eth,
            "block_number":  int(tx["blockNumber"]),
            "block_time":    datetime.fromtimestamp(int(tx["timeStamp"])),
        })
    return flows


async def fetch_internal_transfers(wallet: str, chain: str) -> list[dict]:
    """Fetch internal (contract-to-contract) ETH transfers."""
    cfg = CHAINS[chain]
    data = await _fetch(cfg["explorer"], _params(cfg, {
        "module":  "account",
        "action":  "txlistinternal",
        "address": wallet,
        "sort":    "asc",
    }))

    if data.get("status") != "1":
        return []

    flows = []
    for tx in data["result"]:
        value_eth = int(tx.get("value", 0)) / 1e18
        if value_eth == 0:
            continue
        flows.append({
            "tx_hash":       tx["hash"],
            "chain":         chain,
            "from_address":  tx["from"].lower(),
            "to_address":    tx["to"].lower(),
            "token":         CHAINS[chain]["native"] + "_internal",
            "token_address": "native_internal",
            "amount":        value_eth,
            "block_number":  int(tx["blockNumber"]),
            "block_time":    datetime.fromtimestamp(int(tx["timeStamp"])),
        })
    return flows


async def fetch_nft_transfers(wallet: str, chain: str) -> list[dict]:
    """Fetch ERC-721 NFT transfers for a wallet on a chain."""
    cfg = CHAINS[chain]
    try:
        data = await _fetch(cfg["explorer"], _params(cfg, {
            "module":  "account",
            "action":  "tokennfttx",
            "address": wallet,
            "sort":    "asc",
        }))
    except Exception:
        return []  # NFT endpoint not supported on this explorer

    if data.get("status") != "1":
        return []

    flows = []
    for tx in data["result"]:
        flows.append({
            "tx_hash":       tx["hash"],
            "chain":         chain,
            "from_address":  tx["from"].lower(),
            "to_address":    tx["to"].lower(),
            "token":         f"{tx.get('tokenName','NFT')} #{tx.get('tokenID','?')} [{tx.get('tokenSymbol','')}]",
            "token_address": tx["contractAddress"].lower(),
            "amount":        1.0,
            "block_number":  int(tx["blockNumber"]),
            "block_time":    datetime.fromtimestamp(int(tx["timeStamp"])),
        })
    return flows


async def investigate_address(wallet: str, chain: str) -> dict:
    """
    Full investigation of a wallet on a single chain.
    Returns inflows, outflows, burns, and all raw flows (ERC-20 + native + NFT).
    """
    wallet = wallet.lower()
    token_flows, native_flows, nft_flows = await asyncio.gather(
        fetch_token_transfers(wallet, chain),
        fetch_native_transfers(wallet, chain),
        fetch_nft_transfers(wallet, chain),
    )
    all_flows = token_flows + native_flows + nft_flows
    burn_addr = BURN_ADDRESS.lower()

    return {
        "wallet":    wallet,
        "chain":     chain,
        "all_flows": all_flows,
        "inflows":   [f for f in all_flows if f["to_address"] == wallet],
        "outflows":  [f for f in all_flows if f["from_address"] == wallet],
        "burns":     [f for f in token_flows if f["to_address"] == burn_addr],
        "nfts":      nft_flows,
    }


async def investigate_all_chains(wallet: str) -> dict:
    """
    Investigate a wallet across all supported chains.
    Etherscan V2 chains run sequentially (rate limit: 5 req/s free tier).
    Blockscout/other chains run in parallel alongside them.
    """
    supported = [c for c, cfg in CHAINS.items() if cfg.get("explorer_supported", True)]
    v2_chains    = [c for c in supported if CHAINS[c].get("explorer_v2", True)]
    other_chains = [c for c in supported if not CHAINS[c].get("explorer_v2", True)]

    combined: dict = {"wallet": wallet, "chains": {}}

    async def run_chain(chain, delay=0):
        if delay:
            await asyncio.sleep(delay)
        try:
            r = await investigate_address(wallet, chain)
            combined["chains"][chain] = r
        except Exception as e:
            print(f"[etherscan] {chain} error: {e}")

    # Etherscan V2 chains: stagger by 0.6s each to stay under 5 req/s
    v2_tasks = [run_chain(c, delay=i * 0.6) for i, c in enumerate(v2_chains)]
    # Blockscout/other chains: run all in parallel (generous limits)
    other_tasks = [run_chain(c) for c in other_chains]

    await asyncio.gather(*v2_tasks, *other_tasks)
    return combined

File: src/lib/etherscan.ts
import { CHAINS, BURN_ADDRESS } from "./config";
import { queryRows } from "./db";

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

// Simple semaphore / lock to enforce 300ms gap between Etherscan V2 calls
let lastV2CallTime = 0;
async function throttleV2() {
  const now = Date.now();
  const diff = now - lastV2CallTime;
  if (diff < 350) {
    await sleep(350 - diff);
  }
  lastV2CallTime = Date.now();
}

async function fetchExplorer(url: string, params: Record<string, string>): Promise<any> {
  const queryStr = new URLSearchParams(params).toString();
  const fullUrl = `${url}?${queryStr}`;

  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const response = await fetch(fullUrl, { method: "GET" });
      if (!response.ok) {
        throw new Error(`Explorer returned HTTP status ${response.status}`);
      }
      return await response.json();
    } catch (e) {
      if (attempt === 2) throw e;
      await sleep(Math.pow(1.5, attempt) * 1000);
    }
  }
}

function buildParams(cfg: any, base: Record<string, string>): Record<string, string> {
  const params = { ...base };
  if (cfg.explorer_v2) {
    params["chainid"] = String(cfg.chain_id);
  }
  if (cfg.explorer_key) {
    params["apikey"] = cfg.explorer_key;
  }
  return params;
}

export interface FlowRecord {
  tx_hash: string;
  chain: string;
  from_address: string;
  to_address: string;
  token: string;
  token_address: string;
  amount: number;
  block_number: number;
  block_time: string;
}

export interface InvestigationResult {
  wallet: string;
  chain: string;
  all_flows: FlowRecord[];
  inflows: FlowRecord[];
  outflows: FlowRecord[];
  burns: FlowRecord[];
  nfts: FlowRecord[];
}

export async function fetchTokenTransfers(wallet: string, chain: string): Promise<FlowRecord[]> {
  const cfg = CHAINS[chain];
  if (!cfg || !cfg.explorer_supported) return [];

  if (cfg.explorer_v2) await throttleV2();

  try {
    const data = await fetchExplorer(
      cfg.explorer,
      buildParams(cfg, {
        module: "account",
        action: "tokentx",
        address: wallet,
        sort: "asc",
      })
    );

    if (data.status !== "1" || !Array.isArray(data.result)) {
      return [];
    }

    return data.result.map((tx: any) => {
      const decimals = parseInt(tx.tokenDecimal || "18");
      return {
        tx_hash: tx.hash,
        chain,
        from_address: tx.from.toLowerCase(),
        to_address: tx.to.toLowerCase(),
        token: tx.tokenSymbol,
        token_address: tx.contractAddress.toLowerCase(),
        amount: parseInt(tx.value) / Math.pow(10, decimals),
        block_number: parseInt(tx.blockNumber),
        block_time: new Date(parseInt(tx.timeStamp) * 1000).toISOString(),
      };
    });
  } catch (e) {
    console.error(`[etherscan] fetchTokenTransfers error on ${chain}:`, e);
    return [];
  }
}

export async function fetchNativeTransfers(wallet: string, chain: string): Promise<FlowRecord[]> {
  const cfg = CHAINS[chain];
  if (!cfg || !cfg.explorer_supported) return [];

  if (cfg.explorer_v2) await throttleV2();

  try {
    const data = await fetchExplorer(
      cfg.explorer,
      buildParams(cfg, {
        module: "account",
        action: "txlist",
        address: wallet,
        sort: "asc",
      })
    );

    if (data.status !== "1" || !Array.isArray(data.result)) {
      return [];
    }

    const flows: FlowRecord[] = [];
    for (const tx of data.result) {
      const valueEth = parseInt(tx.value) / 1e18;
      if (valueEth === 0) continue;

      flows.push({
        tx_hash: tx.hash,
        chain,
        from_address: tx.from.toLowerCase(),
        to_address: tx.to ? tx.to.toLowerCase() : "",
        token: cfg.native,
        token_address: "native",
        amount: valueEth,
        block_number: parseInt(tx.blockNumber),
        block_time: new Date(parseInt(tx.timeStamp) * 1000).toISOString(),
      });
    }
    return flows;
  } catch (e) {
    console.error(`[etherscan] fetchNativeTransfers error on ${chain}:`, e);
    return [];
  }
}

export async function fetchInternalTransfers(wallet: string, chain: string): Promise<FlowRecord[]> {
  const cfg = CHAINS[chain];
  if (!cfg || !cfg.explorer_supported) return [];

  if (cfg.explorer_v2) await throttleV2();

  try {
    const data = await fetchExplorer(
      cfg.explorer,
      buildParams(cfg, {
        module: "account",
        action: "txlistinternal",
        address: wallet,
        sort: "asc",
      })
    );

    if (data.status !== "1" || !Array.isArray(data.result)) {
      return [];
    }

    const flows: FlowRecord[] = [];
    for (const tx of data.result) {
      const valueEth = parseInt(tx.value || "0") / 1e18;
      if (valueEth === 0) continue;

      flows.push({
        tx_hash: tx.hash,
        chain,
        from_address: tx.from.toLowerCase(),
        to_address: tx.to.toLowerCase(),
        token: `${cfg.native}_internal`,
        token_address: "native_internal",
        amount: valueEth,
        block_number: parseInt(tx.blockNumber),
        block_time: new Date(parseInt(tx.timeStamp) * 1000).toISOString(),
      });
    }
    return flows;
  } catch (e) {
    console.error(`[etherscan] fetchInternalTransfers error on ${chain}:`, e);
    return [];
  }
}

export async function fetchNftTransfers(wallet: string, chain: string): Promise<FlowRecord[]> {
  const cfg = CHAINS[chain];
  if (!cfg || !cfg.explorer_supported) return [];

  if (cfg.explorer_v2) await throttleV2();

  try {
    const data = await fetchExplorer(
      cfg.explorer,
      buildParams(cfg, {
        module: "account",
        action: "tokennfttx",
        address: wallet,
        sort: "asc",
      })
    );

    if (data.status !== "1" || !Array.isArray(data.result)) {
      return [];
    }

    return data.result.map((tx: any) => ({
      tx_hash: tx.hash,
      chain,
      from_address: tx.from.toLowerCase(),
      to_address: tx.to.toLowerCase(),
      token: `${tx.tokenName || "NFT"} #${tx.tokenID || "?"} [${tx.tokenSymbol || ""}]`,
      token_address: tx.contractAddress.toLowerCase(),
      amount: 1.0,
      block_number: parseInt(tx.blockNumber),
      block_time: new Date(parseInt(tx.timeStamp) * 1000).toISOString(),
    }));
  } catch (e) {
    // Some networks/explorers do not support token nft endpoint
    return [];
  }
}

export async function investigateAddress(wallet: string, chain: string): Promise<InvestigationResult> {
  const cleanWallet = wallet.toLowerCase();
  const burnAddr = BURN_ADDRESS.toLowerCase();

  // Try to query the database cache first
  try {
    const cachedFlows = await queryRows(
      `SELECT tx_hash, chain, from_address, to_address, token, token_address, amount, block_number, block_time 
       FROM address_flows 
       WHERE (from_address = ? OR to_address = ?) AND chain = ?`,
      [cleanWallet, cleanWallet, chain]
    );

    if (cachedFlows && cachedFlows.length > 0) {
      const flows: FlowRecord[] = cachedFlows.map((row: any) => ({
        tx_hash: row.tx_hash,
        chain: row.chain,
        from_address: row.from_address.toLowerCase(),
        to_address: row.to_address ? row.to_address.toLowerCase() : "",
        token: row.token,
        token_address: row.token_address ? row.token_address.toLowerCase() : "",
        amount: Number(row.amount),
        block_number: Number(row.block_number),
        block_time: row.block_time,
      }));

      const tokenFlows = flows.filter((f) => f.token_address !== "native" && f.token_address !== "native_internal");
      const nftFlows = flows.filter((f) => f.token.includes("#"));

      return {
        wallet: cleanWallet,
        chain,
        all_flows: flows,
        inflows: flows.filter((f) => f.to_address === cleanWallet),
        outflows: flows.filter((f) => f.from_address === cleanWallet),
        burns: tokenFlows.filter((f) => f.to_address === burnAddr),
        nfts: nftFlows,
      };
    }
  } catch (e) {
    console.warn(`[etherscan] failed to query database cache for ${cleanWallet}:`, e);
  }

  // Run in parallel if not found in database cache
  const [tokenFlows, nativeFlows, nftFlows] = await Promise.all([
    fetchTokenTransfers(cleanWallet, chain),
    fetchNativeTransfers(cleanWallet, chain),
    fetchNftTransfers(cleanWallet, chain),
  ]);

  const all_flows = [...tokenFlows, ...nativeFlows, ...nftFlows];

  return {
    wallet: cleanWallet,
    chain,
    all_flows,
    inflows: all_flows.filter((f) => f.to_address === cleanWallet),
    outflows: all_flows.filter((f) => f.from_address === cleanWallet),
    burns: tokenFlows.filter((f) => f.to_address === burnAddr),
    nfts: nftFlows,
  };
}

export async function investigateAllChains(wallet: string): Promise<{ wallet: string; chains: Record<string, InvestigationResult> }> {
  const supported = Object.keys(CHAINS).filter((c) => CHAINS[c].explorer_supported);
  const v2Chains = supported.filter((c) => CHAINS[c].explorer_v2);
  const otherChains = supported.filter((c) => !CHAINS[c].explorer_v2);

  const combined: Record<string, InvestigationResult> = {};

  const runChain = async (chain: string, delay = 0) => {
    if (delay) await sleep(delay);
    try {
      combined[chain] = await investigateAddress(wallet, chain);
    } catch (e) {
      console.error(`[etherscan] investigateAllChains ${chain} error:`, e);
    }
  };

  // Stagger Etherscan V2 chains to avoid rate limits
  const v2Promises = v2Chains.map((c, i) => runChain(c, i * 600));
  // Run other chains in parallel
  const otherPromises = otherChains.map((c) => runChain(c));

  await Promise.all([...v2Promises, ...otherPromises]);

  return {
    wallet,
    chains: combined,
  };
}
